src/downloadAzureProject/setupProjectFolder.ts

```diff
@@ -0,0 +1,101 @@
+import { RequestPrepareOptions } from '@azure/ms-rest-js';
+import * as extract from 'extract-zip';
+import * as querystring from 'querystring';
+import * as vscode from 'vscode';
+import { IActionContext } from 'vscode-azureextensionui';
+import { initProjectForVSCode } from '../commands/initProjectForVSCode/initProjectForVSCode';
+import { ProjectLanguage } from '../constants';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import { SlotTreeItemBase } from "../tree/SlotTreeItemBase";
+import { getNameFromId } from '../utils/azure';
+import { requestUtils } from '../utils/requestUtils';
+import { getRequiredQueryParameter, HandleUriActions } from './handleUri';
+
+export async function setupProjectFolder(uri: vscode.Uri, vsCodeFilePathUri: vscode.Uri, context: IActionContext): Promise<void> {
+    const parsedQuery: querystring.ParsedUrlQuery = querystring.parse(uri.query);
+    const resourceId: string = getRequiredQueryParameter(parsedQuery, 'resourceId');
+    const devContainerName: string = getRequiredQueryParameter(parsedQuery, 'devcontainer');
+    const language: string = getRequiredQueryParameter(parsedQuery, 'language');
+    const action: string = getRequiredQueryParameter(parsedQuery, 'action');
+    const downloadAppContent: boolean = action === HandleUriActions.downloadContentAndSetupProject;
+
+    const toBeDeletedFolderPathUri: vscode.Uri = vscode.Uri.joinPath(vsCodeFilePathUri, 'temp');
+
+    try {
+        const functionAppName: string = getNameFromId(resourceId);
+        const downloadFilePath: string = vscode.Uri.joinPath(toBeDeletedFolderPathUri, `${functionAppName}.zip`).fsPath;
+
+        await vscode.window.withProgress({ location: vscode.ProgressLocation.Notification, title: localize('settingUpFunctionAppLocalProjInfoMessage', `Setting up project for function app '${functionAppName}' with language '${language}'.`) }, async () => {
+            if (downloadAppContent) {
+                // NOTE: We don't want to download app content for compiled languages.
+                const slotTreeItem: SlotTreeItemBase | undefined = await ext.tree.findTreeItem(resourceId, { ...context, loadAll: true });
+                const hostKeys: WebSiteManagementModels.HostKeys | undefined = await slotTreeItem?.client.listHostKeys();
+                const defaultHostName: string | undefined = slotTreeItem?.client.defaultHostName;
+
+                if (!!hostKeys && hostKeys.masterKey && defaultHostName) {
+                    const requestOptions: RequestPrepareOptions = {
+                        method: 'GET',
+                        url: `https://${defaultHostName}/admin/functions/download?includeCsproj=true&includeAppSettings=true`,
+                        headers: { 'x-functions-key':  hostKeys.masterKey }
+                    };
+                    await requestUtils.downloadFile(requestOptions, downloadFilePath);
+                }
```

Comment:
This should throw an error if we're failing to get the tree item or host keys rather than silently continuing

Reply:
Throwing an error in else